rollout/sync.go

```diff
@@ -619,14 +620,24 @@ func (c *rolloutContext) calculateRolloutConditions(newStatus v1alpha1.RolloutSt
 			msg := fmt.Sprintf(conditions.ReplicaSetCompletedMessage, rsName)
 			progressingCondition := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionTrue, conditions.NewRSAvailableReason, msg)
 			conditions.SetRolloutCondition(&newStatus, *progressingCondition)
-		case conditions.RolloutProgressing(c.rollout, &newStatus):
+		case conditions.RolloutProgressing(c.rollout, &newStatus) || becameIncomplete:
 			// If there is any progress made, continue by not checking if the rollout failed. This
 			// behavior emulates the rolling updater progressDeadline check.
 			msg := fmt.Sprintf(conditions.RolloutProgressingMessage, c.rollout.Name)
 			if c.newRS != nil {
 				msg = fmt.Sprintf(conditions.ReplicaSetProgressingMessage, c.newRS.Name)
 			}
-			condition := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionTrue, conditions.ReplicaSetUpdatedReason, msg)
+			reason := conditions.ReplicaSetUpdatedReason
+
+			// When a fully promoted rollout becomes Incomplete, e.g., due to the ReplicaSet status changes like
+			// pod restarts, evicted -> recreated, we'll need to reset the rollout's condition to `PROGRESSING` to
+			// avoid any timeouts.
+			if becameIncomplete {
```

Comment:
@huikang i have a feeling that this check will cause the Progressing reason to be `ReplicaSetUpdatedReason` after two reconciliations. Which is why in my suggestion i had the check:

```
if newStatus.StableRS == newStatus.CurrentPodHash {
```

Do you agree?